### PR TITLE
Replace strptime with custom implementation

### LINKED ISSUE #5884: [Ruby] Decoding JSON raises ParseError on Windows
**What version of protobuf and what language are you using?**

Version: v3.7.0
Language: Ruby

**What operating system (Linux, Windows, ...) and version?**

Windows 10

**What runtime / compiler are you using (e.g., python version or gcc version)**

Ruby 2.6.1

**What did you do?**

Install latest google-cloud-pubsub and google-protobuf gems, and run this code:

```ruby
require "google/cloud/pubsub/v1"

now = Time.now
now_ts = Google::Protobuf::Timestamp.new seconds: now.to_i, nanos: now.nsec
msg = Google::Cloud::PubSub::V1::PubsubMessage.new publish_time: now_ts

msg.to_json #=> "{\"attributes\":{},\"publishTime\":\"2019-03-13T15:40:43.729054800Z\"}"

decoded_msg = Google::Cloud::PubSub::V1::PubsubMessage.decode_json msg.to_json
#=> Traceback (most recent call last):
#           1: from pubsub.rb:9:in `<main>'
#   pubsub.rb:9:in `decode_json': Error occurred during parsing: error parsing timestamp: mingw doesn't support strptime (Google::Protobuf::ParseError)
```

**What did you expect to see**

The ability to create a new protobuf resource object by decoding JSON that contains a timestamp.

**What did you see instead?**

A `Google::Protobuf::ParseError` error.

**Anything else we should know about your project / environment**

This appears to only happen on Windows, and [it looks intentional](https://github.com/protocolbuffers/protobuf/blob/v3.7.0/ruby/ext/google/protobuf_c/upb.c#L14125-L14129). But Ruby will parse the timestamp string without issue, so I don't see why Windows couldn't use `Time.parse` instead of `strptime`.

```ruby
require "time"

time_obj = Time.parse "2019-03-13T15:40:43.729054800Z"
timestamp_obj = Google::Protobuf::Timestamp.new seconds: time_obj.to_i, nanos: time_obj.nsec
```

### PR DESCRIPTION
**Pull request by TeBoring**
Fix #5884

### CODE REVIEW

File: ruby/ext/google/protobuf_c/upb.c [hunk 1]
```diff
@@ -14106,49 +14128,100 @@ static bool end_duration_base(upb_json_parser *p, const char *ptr) {
...
+  const char *buf;
+  char *end;
+  int val;
+
+  /* atoi() and friends unfortunately do not support specifying the length of
+   * the input string, so we need to force a copy into a NULL-terminated buffer. */
+  multipart_text(p, "\0", 1, false);
+
+  buf = accumulate_getptr(p, &len);
+  val = atoi(buf);
```
Comment (haberman) [inline ruby/ext/google/protobuf_c/upb.c]: please use strtol() instead so you can check for errors.

Comment (TeBoring) [reply ruby/ext/google/protobuf_c/upb.c]: strtol cannot handle leading 0. It parses 08 as 0.

Comment (haberman) [reply ruby/ext/google/protobuf_c/upb.c]: If our regex is already validating the input, then I'm ok with atoi().

File: ruby/ext/google/protobuf_c/upb.c [hunk 2]
```diff
@@ -14311,7 +14311,7 @@ static bool end_timestamp_zone(upb_json_parser *p, const char *ptr) {
   }
 
   /* Normalize tm */
-  seconds = timegm(&p->tm);
+  seconds = mktime(&p->tm) - timezone;
```
Comment (knu) [inline ruby/ext/google/protobuf_c/upb.c]: This does not work on *BSD platforms where `timezone` is not a long int variable but a [function](https://www.freebsd.org/cgi/man.cgi?query=timezone).

Is this change due to timegm() not available everywhere?

**Pull request merged**